keras_cv/training/simclr/simclr.py

```diff
@@ -0,0 +1,175 @@
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+
+from keras_cv.layers import preprocessing
+from keras_cv.losses import SimCLRLoss
+
+
+class SimCLR(keras.Model):
+    def __init__(
+        self,
+        encoder,
+        include_probing,
+        classes=None,
+        projection_width=128,
+        augmenter=None,
+        include_rescaling=None,
+    ):
+        super().__init__()
+
+        self.include_probing = include_probing
+        self.projection_width = projection_width
+
+        if not augmenter and not include_rescaling:
+            raise ValueError(
+                "`include_rescaling` is required when using the default augmenter."
+            )
+        self.augmenter = augmenter or preprocessing.Augmenter(
+            [
+                preprocessing.RandomFlip("horizontal"),
+                preprocessing.RandomTranslation(0.25, 0.25),
+                preprocessing.RandomZoom((-0.5, 0.0), (-0.5, 0.0)),
+                preprocessing.RandomColorJitter(
+                    value_range=[0, 255] if include_rescaling else [0, 1],
+                    brightness_factor=0.5,
+                    contrast_factor=0.5,
+                    saturation_factor=(0.3, 0.7),
+                    hue_factor=0.5,
+                ),
+            ]
+        )
+
+        self.encoder = encoder
+
+        self.projection_top = keras.Sequential(
+            [
+                keras.Input(shape=(self.encoder.output.shape[-1],)),
+                layers.Dense(self.projection_width, activation="relu"),
+                layers.Dense(self.projection_width),
+            ],
+            name="projection_top",
+        )
+
+        if self.include_probing:
+            if not classes:
+                raise ValueError(
+                    "`classes` must be specified when `include_probing` is `True`."
+                )
+            self.probing_top = keras.Sequential(
+                [
+                    layers.Input(shape=(self.encoder.output.shape[-1],)),
+                    layers.Dense(classes),
+                ],
+                name="linear_probe",
+            )
+
+    def compile(
+        self, contrastive_optimizer, probe_optimizer=None, temperature=0.1, **kwargs
+    ):
+        super().compile(**kwargs)
+
+        # We call the contrastive optimizer `optimizer` so that Keras components
+        # such as the ReduceLROnPlateau callback can correctly update this
+        # optimizer.
+        self.optimizer = contrastive_optimizer
+        self.simclr_loss = SimCLRLoss(temperature)
```

Comment:
lets just require users to pass this as loss:

compile(loss=SimCLRLoss(temperature=0.1))